Calculate payouts correctly

luckysori opened the pull request:
Fixes #1580.

luckysori committed:
Clarify documentation of SETTLEMENT_INTERVAL

As a side-effect it no longer mentions maker and taker.

luckysori committed:
Update OpeningFee doc-string

luckysori committed:
Rename slightly confusing tests

luckysori committed:
Update EventKind doc-string

luckysori committed:
Remove unused method on Dlc

luckysori committed:
Comment on reason for using arbitrary script pubkey

klochowicz commented on model/src/cfd.rs:
```diff
@@ -918,11 +918,15 @@ impl Cfd {
+        anyhow::ensure!(
+            self.role == Role::Maker,
+            "Only the maker can create complete collaborative settlement signing"
+        );
```
I'd go even further by using `assert_debug` or `assert` perhaps - it's not a runtime error, it's a bug in our code if this ever happens, therefore I'd avoid gracefully handling this.

I'm happy also to settle on this one for the time being as it's a giant improvement guarding us from actually doing it when we shouldn't

luckysori replied:
I think you're right. I went for `debug_assert`.

klochowicz commented on daemon/src/maker_cfd.rs:
```diff
@@ -271,8 +271,13 @@ where
 
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd(order_id, &mut conn).await?;
+        let position = cfd.position();
```
this would conflict with: #1585 - suggest syncing with @thomaseizinger

thomaseizinger replied:
Yes, lets not do this please.

klochowicz commented on daemon-tests/src/lib.rs:
```diff
@@ -361,10 +361,10 @@ impl Taker {
     }
 
-    pub async fn trigger_rollover(&self, id: OrderId) {
+    pub async fn trigger_rollover(&self, order_id: OrderId, position: Position) {
```
I thought for a while that this is pretty weird API to need to pass the position, but then realised it's just for the tests..... 😌

luckysori replied:
To be fair it's still pretty weird to me, but I couldn't come up with anything better at this stage.

thomaseizinger replied:
I think this might go away if we return the position together with the event from the command handler?

klochowicz reviewed (approved):
nice job! both with cleanups & features 👍

thomaseizinger commented on daemon/src/rollover_taker.rs:
```diff
@@ -128,6 +132,7 @@ impl Actor {
             (self.oracle_pk, announcement),
             rollover_params,
             Role::Taker,
+            self.position,
```
I'd suggest returning the position from `handle_rollover_accepted_taker` if you need it here.

luckysori replied:
Thanks for the suggestion, that makes things a lot easier.

thomaseizinger commented on model/src/cfd.rs:
```diff
@@ -1890,11 +1897,18 @@ pub fn calculate_payouts_long_taker(
+    }
 }
 
 pub fn calculate_payouts_short_taker(
```
Do we still need this function now?

thomaseizinger reviewed (changes requested):
Thanks for bringing over the refactoring commits. One minor change I'd like to see otherwise LGTM!

luckysori committed:
Check role when performing role-specific methods on Cfd

In debug build only because we should not try to handle such an error
gracefully in production. Instead we want to catch it when testing
before deployment.

thomaseizinger commented on daemon/src/rollover_maker.rs:
```diff
@@ -52,6 +53,7 @@ struct RolloverFailed {
 
 pub struct Actor {
     order_id: OrderId,
+    position: Position,
```
This actor should also retrieve the position by returning it from `accept_rollover_proposal`.

luckysori replied:
You're right!!!!

thomaseizinger commented on daemon/src/setup_maker.rs:
```diff
@@ -101,6 +101,7 @@ impl Actor {
             self.build_party_params.clone_channel(),
             self.sign.clone_channel(),
             Role::Maker,
+            self.order.position_maker,
```
To be consistent, you could also get the position from `start_contract_setup` here instead of passing it in through the `order.` Not sure what is cleaner.

luckysori replied:
It's simpler and more consistent that way because we don't have to pass any extra data to the `setup_taker.rs`.

da-kami commented on model/src/cfd.rs:
```diff
@@ -940,7 +952,9 @@ impl Cfd {
+            self.role,
             self.initial_price,
             self.quantity,
             self.long_leverage,
```
was unable to comment on the line below...

We still always hard-code `Leverage::new(SHORT_LEVERAGE)` - this has to be change to `self.short_leverage` if I understand this correctly. 

Note that this is in multiple places!
I'd recommend removing the `SHORT_LEVERAGE` constant. The maker always being locked in to `Leverage::ONE` should already be correctly handled when constructing the `Cfd`.

thomaseizinger replied:
Nice catch!

luckysori replied:
Makes sense! I'll add a commit on top because this is not strictly related to assigning the payouts to the correct parties.

da-kami replied:
Would be great if you could remove the constant `SHORT_LEVERAGE` so we ensure that we don't use it anymore.

luckysori replied:
> Would be great if you could remove the constant `SHORT_LEVERAGE` so we ensure that we don't use it anymore.

Pretty sure I did!

da-kami replied:
Ah, yeah, sorry. Somhow did not see it in the commit 😅

thomaseizinger reviewed (approved):
Cool, happy with that! Thank you :)

luckysori committed:
Do not hard-code short leverage when calling calculate_payouts

luckysori committed:
Calculate payouts correctly based on role and position
